cmd/containerboot/main.go

```diff
@@ -0,0 +1,426 @@
+		}
+	}
+	if cfg.ProxyTo != "" || cfg.Routes != "" {
+		if err := ensureIPForwarding(); err != nil {
+			log.Printf("Failed to enable IP forwarding: %v", err)
+			log.Printf("To run tailscale as a proxy or router container, IP forwarding must be enabled.")
+			if cfg.InKubernetes {
+				log.Fatalf("You can either set the sysctls as a privileged initContainer, or run the tailscale container with privileged=true.")
+			} else {
+				log.Fatalf("You can fix this by running the container with privileged=true, or the equivalent in your container runtime that permits access to sysctls.")
+			}
+		}
+	}
+
+	// Context is used for all setup stuff until we're in steady
+	// state, so that if something is hanging we eventually time out
+	// and crashloop the container.
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	if cfg.InKubernetes && cfg.KubeSecret != "" && cfg.AuthKey == "" {
+		key, err := findKeyInKubeSecret(ctx, cfg.KubeSecret)
+		if err != nil {
+			log.Fatalf("Getting authkey from kube secret: %v", err)
+		}
+		if key != "" {
+			log.Print("Using authkey found in kube secret")
+			cfg.AuthKey = key
+		} else {
+			log.Print("No authkey found in kube secret and TS_AUTHKEY not provided, login will be interactive if needed.")
+		}
+	}
+
+	st, err := startAndAuthTailscaled(ctx, cfg)
+	if err != nil {
+		log.Fatalf("failed to bring up tailscale: %v", err)
+	}
+
+	if cfg.ProxyTo != "" {
+		if err := installIPTablesRule(ctx, cfg.ProxyTo, st.TailscaleIPs); err != nil {
+			log.Fatalf("installing proxy rules: %v", err)
+		}
+	}
+	if cfg.KubeSecret != "" {
+		if err := storeDeviceID(ctx, cfg.KubeSecret, string(st.Self.ID)); err != nil {
+			log.Fatalf("storing device ID in kube secret: %v", err)
+		}
+		if cfg.AuthOnce {
+			// We were told to only auth once, so any secret-bound
+			// authkey is no longer needed. We don't strictly need to
+			// wipe it, but it's good hygiene.
+			log.Printf("Deleting authkey from kube secret")
+			if err := deleteAuthKey(ctx, cfg.KubeSecret); err != nil {
+				log.Fatalf("deleting authkey from kube secret: %v", err)
+			}
+		}
+	}
+
+	log.Println("Startup complete, waiting for shutdown signal")
+	// Sleep for ever, tailscaled is handled from a goroutine in
+	// startAndAuthTailscaled.
+	select {}
+}
+
+// startAndAuthTailscaled starts the tailscale daemon and attempts to
+// auth it, according to the settings in cfg. If successful, returns a
+// tailscale.LocalClient connected to the running tailscaled.
+func startAndAuthTailscaled(ctx context.Context, cfg *settings) (*ipnstate.Status, error) {
+	args := tailscaledArgs(cfg)
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, unix.SIGTERM, unix.SIGINT)
+	// tailscaled runs without context, since it needs to persist
+	// beyond the startup timeout in ctx.
+	cmd := exec.Command("tailscaled", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	log.Printf("Starting tailscaled")
```

Comment:
Idle musing: we should look at the Linux process contract and see if we need to set anything in `cmd.SysProcAttr`; I think we want `Setpgid: true` here

Reply:
I'm not following how this matters, for the constrained environment of this container?

Reply:
Huge caveat here: I am not a Unix/POSIX expert.

My understanding of process groups is that when a signal is directed to a process group, the signal is delivered to each process that is a member of the group. By default, when a shell `fork`s a subprocess, I believe that it makes the child process the leader of its own (new) process group. By launching `tailscaled` here ourselves, that isn't true any more, and if something delivers a signal to this process, both `containerboot` and `tailscaled` will be part of the same process group and both will receive the signal.

Since I don't know exactly how container runtimes work w.r.t. signals, I think the least-confusing approach here is to do exactly what you're doing to forward a signal from PID1 to the child `tailscaled`, and additionally make `tailscaled` the leader of its own process group so it doesn't get any signals except what `containerboot` explicitly sends it.

(again: not an expert here by any means)

Reply:
sgtm, done. I think in the container setting it might not matter, but since all the plumbing is there in pid 1 to relay sigterm and do an orderly shutdown, we should make it actually work.